### ISSUE TITLE
Audit: keyboard accessibility

### ISSUE THREAD

**Issue by fchristant** [GoogleChrome/lighthouse#10331]
At our org, we're slowly improving our accessibility scores across a vast web property. We're using multiple tools, automated and manual, and LH's current accessibility checks have proven quite useful. 

However, as we know, it does not cover the full spectrum of accessibility testing. One part lacking is auditing for keyboard accessibility. This is a **very** laboursome task. We currently manually check many scenarios by hand. This idea is to explore if perhaps some of it can be automated into audits, both to save time but also to raise awareness to developers.

To be honest, I have low confidence in the feasibility of this, yet it doesn't hurt to try. I'll set out this idea by listing the typical problems we find in our manual audits:

**An element that should be focusable, isn't focusable**
For example, a developer using a div with an onclick, instead of an actual anchor. I don't expect this to be auditable, because you cannot reliably detect the developer's intend.

**An element that is focusable, does not communicate this with visual state**
The typical "remove all outlines because we find them ugly" situation. This can be auditable, but may not be easy. You can't just check for the absence of an outline, as a developer may style the focus state in an alternative way that would still be valid. 

**A pseudo-focusable element that cannot be activated**
Using tab-index="0" to make a non-interactive element focusable. Yet when hitting [enter] in the focus state, nothing happens. There's only a click handler, not a keyup handler. This may be auditable, yet probably not easy, it requires deep JS introspection.

**Off-screen**
A bit rare but a source of weird bugs: developers not hiding things by taking them out of the DOM, instead putting them off-screen. If said element is focusable, this leads to weird issues. Very much an edge-case, probably not worthwhile to even try to audit.

After writing this, I think my confidence in feasibility is now even lower, but still wanted to share it as food for thought.

**Comment by connorjclark** [comment]
cc @johnemau

**Comment by Malvoz** [comment]
@fchristant I support this idea. Can we re-open this?

Particularly, setting `* { outline: 0; }` is still common among developers. An alternative approach that does not negatively affect keyboard-only users is to use `:focus:not(:focus-visible) { outline: 0; }` instead. It'd be great if such a recommendation made its way into Lighthouse.

**Comment by fchristant** [comment]
Thanks for re-opening. I think I closed it back then because of a lack of interest but more importantly the difficulty to automate the auditing of the particular examples I shared. I could totally be wrong about this feasibility, of course. Furthermore, feel free to add other potential keyboard accessibility checks not yet covered, its an issue pretty fluid in scope.

**Comment by paulirish** [comment]
I think these ideas are promising, certainly. Of course feasibility is a concern and I appreciate you thinking through that.

The next step would probably be bringing this proposal to https://github.com/dequelabs/axe-core/. (We don't have any of our own a11y audits, and instead defer entirely to them).
